extensions/amp-ad-network-base/0.1/amp-ad-type-defs.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @typedef {{
```

Comment:
can we type define validator & renderer functions

Reply:
I have these defined in the amp-render.js file, where the renderers and validators are. You can see this in my [other PR](https://github.com/ampproject/amphtml/pull/13582/files#diff-a1de641dbd4a1be139250792916dd6eeR52). I envision these as abstract classes with a single method (render()/validate()).

Reply:
As I previously mentioned, it's normally good idea to have interface and impl in different files, so that import the interface will not import all the impls (though closure compiler might do the job via DCE).

So let's move all interfaces into a separate file, and get included in this PR.

Reply:
Done.